File: backend/analysis/download.py
#!/usr/bin/env python3
"""
Downloads missing NZ public access area layers from ArcGIS FeatureServer.
Idempotent — skips layers that already exist in --data-dir.

Usage: python download.py --data-dir <dir>
"""
import argparse
import json
import sys
import time
from pathlib import Path

import requests

BASE_URL = (
    "https://services2.arcgis.com/b5ADKIcWivL5vNaV/arcgis/rest/services"
    "/Public_Access_Areas/FeatureServer"
)
PAGE_SIZE = 2000
LAYERS = [
    (2, "Easements"),
    (3, "Reserve_Land"),
    (4, "Public_Access_Conservation_Land"),
    (5, "Other_Parks_and_Reserves"),
    (6, "Other_Public_Access_Areas"),
]


def get_count(layer_id: int) -> int:
    r = requests.get(
        f"{BASE_URL}/{layer_id}/query",
        params={"where": "1=1", "returnCountOnly": "true", "f": "json"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["count"]


def download_layer(layer_id: int, name: str) -> dict:
    print(f"Downloading {name}…", file=sys.stderr)
    total = get_count(layer_id)
    print(f"  {total} features", file=sys.stderr)
    features, crs, offset = [], None, 0
    while offset < total:
        r = requests.get(
            f"{BASE_URL}/{layer_id}/query",
            params={
                "where": "1=1", "outFields": "*", "returnGeometry": "true",
                "resultOffset": offset, "resultRecordCount": PAGE_SIZE,
                "f": "geojson", "outSR": "4326",
            },
            timeout=60,
        )
        r.raise_for_status()
        data = r.json()
        batch = data.get("features", [])
        features.extend(batch)
        if crs is None:
            crs = data.get("crs")
        offset += len(batch)
        print(f"  {min(offset, total)}/{total}", file=sys.stderr)
        if not batch or offset >= total:
            break
        time.sleep(0.2)
    return {"type": "FeatureCollection", "crs": crs, "features": features}


def ensure_cache(data_dir: Path) -> None:
    data_dir.mkdir(parents=True, exist_ok=True)
    missing = [(lid, n) for lid, n in LAYERS if not (data_dir / f"{n}.geojson").exists()]
    if not missing:
        return
    for layer_id, name in missing:
        data = download_layer(layer_id, name)
        out = data_dir / f"{name}.geojson"
        with open(out, "w", encoding="utf-8") as f:
            json.dump(data, f)
        print(f"Saved {out}", file=sys.stderr)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--data-dir", required=True)
    ensure_cache(Path(parser.parse_args().data_dir))

File: backend/src/analysis/download.ts
import { existsSync, mkdirSync, writeFileSync } from 'fs'
import path from 'path'

const BASE_URL =
  'https://services2.arcgis.com/b5ADKIcWivL5vNaV/arcgis/rest/services/Public_Access_Areas/FeatureServer'

const PAGE_SIZE = 2000

const LAYERS: [number, string][] = [
  [2, 'Easements'],
  [3, 'Reserve_Land'],
  [4, 'Public_Access_Conservation_Land'],
  [5, 'Other_Parks_and_Reserves'],
  [6, 'Other_Public_Access_Areas'],
]

async function getCount(layerId: number): Promise<number> {
  const url = new URL(`${BASE_URL}/${layerId}/query`)
  url.searchParams.set('where', '1=1')
  url.searchParams.set('returnCountOnly', 'true')
  url.searchParams.set('f', 'json')
  const res = await fetch(url.toString())
  if (!res.ok) throw new Error(`Count request failed: ${res.status}`)
  const data = (await res.json()) as { count: number }
  return data.count
}

async function downloadLayer(layerId: number, name: string): Promise<object> {
  console.error(`Downloading ${name}…`)
  const total = await getCount(layerId)
  console.error(`  ${total} features`)

  const features: object[] = []
  let crs: object | null = null
  let offset = 0

  while (offset < total) {
    const url = new URL(`${BASE_URL}/${layerId}/query`)
    url.searchParams.set('where', '1=1')
    url.searchParams.set('outFields', '*')
    url.searchParams.set('returnGeometry', 'true')
    url.searchParams.set('resultOffset', String(offset))
    url.searchParams.set('resultRecordCount', String(PAGE_SIZE))
    url.searchParams.set('f', 'geojson')
    url.searchParams.set('outSR', '4326')

    const res = await fetch(url.toString(), { signal: AbortSignal.timeout(60_000) })
    if (!res.ok) throw new Error(`Layer request failed: ${res.status}`)
    const data = (await res.json()) as { features?: object[]; crs?: object }

    const batch = data.features ?? []
    features.push(...batch)
    if (crs === null && data.crs) crs = data.crs
    offset += batch.length
    console.error(`  ${Math.min(offset, total)}/${total}`)

    if (!batch.length || offset >= total) break
    await new Promise(r => setTimeout(r, 200))
  }

  return { type: 'FeatureCollection', crs, features }
}

export async function ensureCache(dataDir: string): Promise<void> {
  mkdirSync(dataDir, { recursive: true })
  const missing = LAYERS.filter(([, name]) => !existsSync(path.join(dataDir, `${name}.geojson`)))
  if (!missing.length) return

  for (const [layerId, name] of missing) {
    const data = await downloadLayer(layerId, name)
    const out = path.join(dataDir, `${name}.geojson`)
    writeFileSync(out, JSON.stringify(data))
    console.error(`Saved ${out}`)
  }
}
